distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyParser.java

```diff
@@ -0,0 +1,215 @@
+                throw newPolicyParserException(
+                    policyParser.getTokenLocation(),
+                    policyName,
+                    scopeName,
+                    "expected object [" + ENTITLEMENTS_PARSEFIELD.getPreferredName() + "]"
+                );
+            }
+            if (policyParser.nextToken() != XContentParser.Token.START_ARRAY) {
+                throw newPolicyParserException(
+                    policyParser.getTokenLocation(),
+                    policyName,
+                    scopeName,
+                    "expected array of <entitlement type>"
+                );
+            }
+            List<Entitlement> entitlements = new ArrayList<>();
+            while (policyParser.nextToken() != XContentParser.Token.END_ARRAY) {
+                if (policyParser.currentToken() != XContentParser.Token.START_OBJECT) {
+                    throw newPolicyParserException(
+                        policyParser.getTokenLocation(),
+                        policyName,
+                        scopeName,
+                        "expected object <entitlement type>"
+                    );
+                }
+                if (policyParser.nextToken() != XContentParser.Token.FIELD_NAME) {
+                    throw newPolicyParserException(
+                        policyParser.getTokenLocation(),
+                        policyName,
+                        scopeName,
+                        "expected object <entitlement type>"
+                    );
+                }
+                String entitlementType = policyParser.currentName();
+                Entitlement entitlement = parseEntitlement(scopeName, entitlementType);
+                entitlements.add(entitlement);
+                if (policyParser.nextToken() != XContentParser.Token.END_OBJECT) {
+                    throw newPolicyParserException(policyParser.getTokenLocation(), policyName, scopeName, "expected closing object");
+                }
+            }
+            if (policyParser.nextToken() != XContentParser.Token.END_OBJECT) {
+                throw newPolicyParserException(policyParser.getTokenLocation(), policyName, scopeName, "expected closing object");
+            }
+            return new Scope(scopeName, entitlements);
+        } catch (IOException ioe) {
+            throw new UncheckedIOException(ioe);
+        }
+    }
+
+    protected Entitlement parseEntitlement(String scopeName, String entitlementType) throws IOException {
+        Class<?> entitlementClass;
+        try {
+            entitlementClass = Class.forName(
+                entitlementPackageName
+                    + "."
+                    + Character.toUpperCase(entitlementType.charAt(0))
+                    + entitlementType.substring(1)
+                    + "Entitlement"
+            );
+        } catch (ClassNotFoundException cnfe) {
+            throw newPolicyParserException(
+                policyParser.getTokenLocation(),
```

Comment:
nit: The token location and policy name seem to be the same in all calls to these helper exception methods, could those be removed as args?

Reply:
I cleaned this up with some additional helper methods.